scarlet-stream-adapter-coroutines/src/main/java/com/tinder/streamadapter/coroutines/ChannelForwarder.kt

```diff
@@ -0,0 +1,32 @@
+package com.tinder.streamadapter.coroutines
+
+import com.tinder.scarlet.Stream
+import kotlinx.coroutines.channels.BufferOverflow
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.ReceiveChannel
+import kotlinx.coroutines.channels.trySendBlocking
+
+internal class ChannelForwarder<T>(bufferSize: Int) : Stream.Observer<T> {
+    private val channel = Channel<T>(bufferSize, BufferOverflow.DROP_OLDEST)
+    private var disposable: Stream.Disposable? = null
+
+    fun start(stream: Stream<T>): ReceiveChannel<T> {
+        disposable = stream.start(this)
+        return channel
+    }
+
+    override fun onComplete() {
+        channel.close()
+        disposable?.dispose()
+    }
+
+    override fun onError(throwable: Throwable) {
+        channel.close(throwable)
+        disposable?.dispose()
+    }
+
+    override fun onNext(data: T) {
+        channel.trySendBlocking(data)
+            .exceptionOrNull() ?.let { throw it }
```

Comment:
I think so, this would represent a case when the channel is closed, in which case there should be no subscribers anymore. So I think this would be a true failure case

Reply:
Hmm… Have a look at the `ObservableSource.asFlow()` extension from `coroutines-rx2`:
https://github.com/Kotlin/kotlinx.coroutines/blob/15815923970b5da2e9a1bd0d68293869341714c1/reactive/kotlinx-coroutines-rx2/src/RxConvert.kt#L80-L90

It looks like the exception gets swallowed in the case that the channel is closed but the `Observable` still tries to emit a value.

Reply:
Maybe I am misunderstanding, but the referenced code only traps a InterruptedException which it expected to be thrown. All other exceptions get propagated upwards. Flows maintain exception transparency with the caller. So the onError would then be triggered and the stream would be closed

Reply:
58b07ed I see what you are saying now. Sorry long day, the trySendBlocking will return a result in that referenced code and is just not handled for errors. I am ok with this with this precedence